File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    // time O(2^n)
    // gerateing
    void gen(int k, int n, vector<vector<int>>& res, vector<int> temp, vector<int>& nums){
        if(k==n){
            // goal
            res.push_back(temp);
        }else{
            //choice - no constraint
            gen(k+1,n,res,temp,nums);
            temp.push_back(nums[k]);
            gen(k+1,n,res,temp,nums);
        }
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> temp;
        vector<vector<int>> res;
        int n = nums.size(), k = 0;
        gen(k,n,res,temp,nums);
        return res;
    }
};

File: 0078-subsets/0078-subsets.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function(nums) {
    const n = nums.length;
    const res = [];
    const temp = [];
    const gen = (k)=>{
        if(k===n) res.push(Array.from(temp));
        else{
            gen(k+1);
            temp.push(nums[k]);
            gen(k+1);
            temp.pop();
        }
    }
    gen(0);
    return res;
};